### PR TITLE
feat(m3): notebook hygiene fixes for classification/regression empty-charts

### PR DESCRIPTION
**Pull request by JuanCamiloDorado**
## Resumen

Cierra el síntoma recurrente del notebook generado: gráficos vacíos en clasificación binaria (feature_importance todo en 0, matriz de confusión 1×1) que se observó hoy en QA del end-product.

## Diagnóstico

Origen real (no cosmético):

1. El split temporal podía dejar **una sola clase** en `y_train` o `y_test`.
2. sklearn emitía `UndefinedMetricWarning("Only one class present in y_true")`.
3. La señal quedaba **silenciada globalmente** por `warnings.filterwarnings("ignore")` en el preámbulo.
4. Sin AUC-ROC/AUC-PR, las métricas reportadas (accuracy, classification_report) **disfrazaban** un fit donde el modelo predecía una sola clase con accuracy "perfecto" sobre un test colapsado.
5. Los gráficos posteriores se renderizaban con `feature_importances_ = [0, 0, 0, ...]` o matrices `[[40, 0]]`.

## Cinco fixes transversales (LLM-instruction layer, sin cambios en Python)

Aplicados en `M3_NOTEBOOK_ALGO_PROMPT` y `M3_NOTEBOOK_BASE_TEMPLATE`:

### 1. Defensa extra anti-leakage por naming
Aun con contrato, excluir features cuyo nombre coincida con patrones temporales-posteriores (`retention_m*`, `days_to_*`, `cancellation*`, `_post_*`, `m3_/m6_/m12_`) salvo que sean el target.

### 2. Métricas obligatorias para binario (Regla H reforzada)
- **AUC-ROC + AUC-PR obligatorios** vía `predict_proba` cuando `y_test.nunique() == 2` (las únicas métricas que delatan un predictor de clase mayoritaria).
- **Pesos de clase obligatorios:** `class_weight="balanced"` para LR/RF/SVC; `scale_pos_weight = (y_train==0).sum() / (y_train==1).sum()` para XGBoost.
- Imprimir `y_train.value_counts(normalize=True)` antes del fit.
- MAE añadido a regresión.

### 3. Guarda post-split obligatoria (Regla I reforzada)
```python
if y_train.nunique() < 2 or y_test.nunique() < 2:
    print("⚠️ SPLIT DEGENERADO ...")
    model = None
    # salir, NO ejecutar .fit() ni los plots
```
Las celdas 2b/2c/2d deben respetar `if model is None`.

### 4. Higiene de feature_cols (nueva Regla K-bis)
Receta determinista antes de construir `X`:
- Numéricas + categóricas con cardinalidad ≤ 20.
- Drop ID-like (cardinalidad == n_filas o nombre con "id").
- Drop near-constants (`nunique <= 1`) y high-null (>50% NaN).
- One-hot via `pd.get_dummies(..., drop_first=True, dummy_na=False)`.
- `assert X.shape[1] >= 1` o saltar el algoritmo.

Elimina la clase de bug `feature_importance` todo en 0.

### 5. Carve-out al try/except (Regla 6 ampliada)
Las guardas explícitas de pre-fit (split degenerado, feature_cols vacío, target ausente) **no deben** quedar tragadas por `except Exception`. Su `print("⚠️ ...")` debe ser visible.

### Bonus en BASE_TEMPLATE
`warnings.filterwarnings("ignore")` → `"default"`. Solo `DeprecationWarning` se sigue silenciando. Ahora `UndefinedMetricWarning` aflora.

## Tests

- Nuevos: `backend/tests/test_m3_notebook_classification_hygiene.py` (7 tests, cero LLM/red/DB).
- Smoke de `.format(**)` con todas las claves esperadas para detectar `{}` mal escapado.

## Validación

- ✅ `uv run --directory backend pytest -q` → **446 passed, 12 skipped**
- ✅ `uv run --directory backend mypy src` → **Success: no issues found in 44 source files**
- Frontend no tocado (sin cambios en `frontend/`).

## Scope explícito

- Es una intervención **transversal** (un solo prompt, todos los algoritmos heredan).
- **NO** introduce templates por familia (clasificación / regresión / time-series / clustering / recommendation / NLP). Ese trabajo es PR B y se trackeará en una issue separada.
- **NO** reduce el catálogo de algoritmos (Issue #230 sigue intacto).

## Sensitive area

Toca `backend/src/case_generator/prompts.py`. Todos los cambios son aditivos sobre reglas existentes; no se altera la orquestación del grafo ni los contratos `studentProfile` / `M1..M6`.

### CODE REVIEW

**Comment by JuanCamiloDorado** [comment]
## Review feedback addressed — commit `1ec5480`

Las cuatro sugerencias del review fueron analizadas y aceptadas como **válidas** (sin falsos positivos). Aplicadas todas en un único fixup commit; suite completa verde.

### 1. AUC silenciosamente saltado para SVC y roto con etiquetas string
- `Regla H` reescrita: ahora deriva `pos_label` de `model.classes_` (fallback al máximo ordenado de `y_train.unique()`), binariza `y_test` contra ese `pos_label` antes de `roc_auc_score` / `average_precision_score`, e incluye **fallback a `decision_function`** cuando `predict_proba` no existe (caso por defecto de `SVC`). Si ninguno de los dos está disponible, imprime un skip explícito en lugar de fallar silenciosamente.
- Documentado: `SVC` requiere `probability=True` si se quiere AUC vía proba.

### 2. `scale_pos_weight` asumía etiquetas numéricas 0/1
- Reescrito a `vc = y_train.value_counts(); scale_pos_weight = float(vc.max()) / float(max(vc.min(), 1))`. Funciona para cualquier binaria string/int y evita división por cero.

### 3. Bloque `K-bis` usaba fence ```` ``` ```` (contradice Regla absoluta 3)
- Convertido a lista numerada con líneas inline en backticks + advertencia explícita: *“NO emitas estos pasos como bloque cercado con triple backtick — emítelos como código Python normal de la celda”*. La receta (drop ID-like, near-constants, high-null, one-hot vía `pd.get_dummies`, assert `X.shape[1] >= 1`) se preserva intacta.

### 4. Smoke test tautológico
- El antiguo `assert "{" not in ... or "{{" not in rendered` siempre pasaba porque tras `format()` `{{` se vuelve `{`, así que el operando derecho era siempre `True`.
- Reemplazado por aserciones positivas: ningún placeholder conocido sobrevive (`{m3_content}`, `{algoritmos}`, `{familias_meta}`, `{case_title}`, `{output_language}`, `{dataset_contract_block}`, `{data_gap_warnings_block}`), todos los sentinelas inyectados aparecen, y el floor de longitud es razonable.

### Validación
- `uv run --directory backend pytest -q` → **482 passed, 12 skipped** (≈116 s)
- `uv run --directory backend mypy src` → **Success: no issues found in 44 source files**

Listo para resolver los review threads.
